BUG: if zero derivative at root, then Newton fails with RuntimeWarning

As indicated in [this discussion](../pull/8357#discussion_r192612946) by @pvanmulbregt

>... if the derivative is 0 at the root? E.g. `f(x)=x**3 - x**2`, starting at `x0=0` (immediately at the root), ... It looks like this might be considered a failure


### Reproducing code example:
```
from scipy import optimize

# use secant
r = optimize.newton(lambda x: x**3 - x**2, x0=0)
# -1.3552527156068808e-20  close enough to zero to be the correct answer

# now try Newton
r = optimize.newton(lambda x: x**3 - x**2, x0=0, fprime=lambda x: 3*x**2 - 2*x)
# raises RuntimeWarning but solution is zero

# or Halley's
r = optimize.newton(lambda x: x**3 - x**2, x0=0, fprime=lambda x: 3*x**2 - 2*x, fprime2=lambda x: 6*x - 2)
# also raises RuntimeWarning even though solution should be zero
```

### Error message:
```
~/miniconda/lib/python3.6/site-packages/scipy/optimize/zeros.py:169: RuntimeWarning: derivative was zero.
  warnings.warn(msg, RuntimeWarning)
```

### Scipy/Numpy/Python version information:
```
1.1.0 1.14.3 sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)
```

So I can fix this issue by testing `fval` first:

```diff
diff --git a/scipy/optimize/zeros.py b/scipy/optimize/zeros.py
index 7272ff0fb..eb7c9fdea 100644
--- a/scipy/optimize/zeros.py
+++ b/scipy/optimize/zeros.py
@@ -187,6 +187,9 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
     # so it still works if x0 is complex.
     p0 = 1.0 * x0
+    # check if initial guess is the solution
+    if func(p0, *args) == 0:
+        return _results_select(full_output, (p0, 0, 0, _ECONVERGED))
     funcalls = 0
     if fprime is not None:
         # Newton-Raphson method
```

... but then it causes [`scipy/optimize/tests/test_zeros.py::TestBasic::test_deriv_zero_warning`](../tree/2ce7116d2/scipy/optimize/tests/test_zeros.py#L99-L102) to fail. _NB_ `test_deriv_zero_warning` was written 8 years ago by @WarrenWeckesser  and explicitly tests the condition when `fval` and `fder` are both zero, which we are now considering to change to not raising `RuntimeWarning`.  So now I'm inclined to wait on #8904 and not block #8357 in order to allow the SciPy community to weigh in on whether they want to change when `RuntimeWarning` is raised. [I've raised the question on the SciPy Dev mailing list to get more feedback.](https://mail.python.org/pipermail/scipy-dev/2018-June/022861.html)

Here is the traceback:
<details>

```python
self = <scipy.optimize.tests.test_zeros.TestBasic object at 0x7f2062993898>

    def test_deriv_zero_warning(self):
        func = lambda x: x**2
        dfunc = lambda x: 2*x
>       assert_warns(RuntimeWarning, cc.newton, func, 0.0, dfunc)

dfunc      = <function TestBasic.test_deriv_zero_warning.<locals>.<lambda> at 0x7f20629956a8>
func       = <function TestBasic.test_deriv_zero_warning.<locals>.<lambda> at 0x7f20629958c8>
self       = <scipy.optimize.tests.test_zeros.TestBasic object at 0x7f2062993898>

scipy/optimize/tests/test_zeros.py:102:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../../../venv/lib/python3.5/site-packages/numpy/testing/nose_tools/utils.py:1731: in assert_warns
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <contextlib._GeneratorContextManager object at 0x7f2062993dd8>, type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):
        if type is None:
            try:
>               next(self.gen)
E               AssertionError: No warning raised when calling newton

self       = <contextlib._GeneratorContextManager object at 0x7f2062993dd8>
traceback  = None
type       = None
value      = None

/usr/lib/python3.5/contextlib.py:66: AssertionError
```
</details>

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -97,7 +97,7 @@ class TestBasic(object):
                     x, r = zeros.newton(f1, x0, maxiter=iters, disp=True, **kwargs)
 
     def test_deriv_zero_warning(self):
-        func = lambda x: x**2
+        func = lambda x: x**2 - 2.0
         dfunc = lambda x: 2*x
         assert_warns(RuntimeWarning, cc.newton, func, 0.0, dfunc)
 

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -171,8 +171,9 @@ def test_complex_halley():
     assert_allclose(f(y, *coeffs), 0, atol=1e-6)
 
 
+@pytest.mark.filterwarnings("error")
 def test_gh8904_zeroder_at_root_fails():
-    """Test that Newton or Halley don't fail if zero derivative at root"""
+    """Test that Newton or Halley don't warn if zero derivative at root"""
 
     # a function that has a zero derivative at it's root
     def f_zeroder_root(x):

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -175,13 +175,15 @@ def test_complex_halley():
 def test_gh8904_zeroder_at_root_fails():
     """Test that Newton or Halley don't warn if zero derivative at root"""
 
+    xtol = finfo(float).eps
+
     # a function that has a zero derivative at it's root
     def f_zeroder_root(x):
         return x ** 3 - x ** 2
 
     # should work with secant
     r = zeros.newton(f_zeroder_root, x0=0)
-    assert_allclose(r, 0, atol=1e-16)
+    assert_allclose(r, 0, atol=xtol)
 
     # 1st derivative
     def fder(x):
@@ -189,10 +191,10 @@ def test_gh8904_zeroder_at_root_fails():
 
     # should work with newton and halley
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder)
-    assert_allclose(r, 0, atol=1e-16)
+    assert_allclose(r, 0, atol=xtol)
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder,
                      fprime2=lambda x: 6 * x - 2)
-    assert_allclose(r, 0, atol=1e-16)
+    assert_allclose(r, 0, atol=xtol)
 
     # also test that if a root is found we do not raise RuntimeWarning even if
     # the derivative is zero, EG: at x = 0.5, then fval = -0.125 and
@@ -200,7 +202,7 @@ def test_gh8904_zeroder_at_root_fails():
     # root, but if the solver continued with that guess, then it will calculate
     # a zero derivative, so it should return the root w/o RuntimeWarning
     r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder)
-    assert_allclose(r, 0, atol=1e-16)
+    assert_allclose(r, 0, atol=xtol)
     r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder,
                      fprime2=lambda x: 6 * x - 2)
-    assert_allclose(r, 0, atol=1e-16)
+    assert_allclose(r, 0, atol=xtol)

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -191,7 +191,7 @@ def test_gh8904_zeroder_at_root_fails():
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder)
     assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder,
-                     fprime2=lambda x: 6 * x - 2)
+                     fprime2=lambda x: 6*x - 2)
     assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
 
     # also test that if a root is found we do not raise RuntimeWarning even if

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -169,3 +169,27 @@ def test_complex_halley():
     y = zeros.newton(f, z, args=coeffs, fprime=f_1, fprime2=f_2, tol=1e-6)
     # (-0.75000000000000078+1.1989578808281789j)
     assert_allclose(f(y, *coeffs), 0, atol=1e-6)
+
+
+def test_gh8904_zeroder_at_root_fails():
+    """Test that Newton or Halley don't fail if zero derivative at root"""
+
+    # a function that has a zero derivative at it's root
+    def f_zeroder_root(x):
+        return x ** 3 - x ** 2
+
+    # should work with secant
+    r = zeros.newton(f_zeroder_root, x0=0)
+    assert_allclose(r, 0, atol=1e-16)
+
+    # 1st derivative
+    def fder(x):
+        return 3 * x ** 2 - 2 * x
+
+    # should work with newton and halley
+    r = zeros.newton(f_zeroder_root, x0=0, fprime=fder)
+    assert_allclose(r, 0, atol=1e-16)
+    r = zeros.newton(f_zeroder_root, x0=0, fprime=fder,
+                     fprime2=lambda x: 6 * x - 2)
+    assert_allclose(r, 0, atol=1e-16)
+

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -203,4 +203,4 @@ def test_gh8904_zeroder_at_root_fails():
     assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
     r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder,
                      fprime2=lambda x: 6 * x - 2)
-    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
+    assert_allclose(r, 0, atol=1e-6)

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -201,6 +201,4 @@ def test_gh8904_zeroder_at_root_fails():
     # a zero derivative, so it should return the root w/o RuntimeWarning
     r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder)
     assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
-    r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder,
-                     fprime2=lambda x: 6 * x - 2)
-    assert_allclose(r, 0, atol=1e-6)
+    # doesn't apply to halley

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -194,3 +194,13 @@ def test_gh8904_zeroder_at_root_fails():
                      fprime2=lambda x: 6 * x - 2)
     assert_allclose(r, 0, atol=1e-16)
 
+    # also test that if a root is found we do not raise RuntimeWarning even if
+    # the derivative is zero, EG: at x = 0.5, then fval = -0.125 and
+    # fder = -0.25 so the next guess is 0.5 - (-0.125/-0.5) = 0 which is the
+    # root, but if the solver continued with that guess, then it will calculate
+    # a zero derivative, so it should return the root w/o RuntimeWarning
+    r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder)
+    assert_allclose(r, 0, atol=1e-16)
+    r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder,
+                     fprime2=lambda x: 6 * x - 2)
+    assert_allclose(r, 0, atol=1e-16)

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -171,13 +171,13 @@ def test_complex_halley():
     assert_allclose(f(y, *coeffs), 0, atol=1e-6)
 
 
-@pytest.mark.filterwarnings("error")
+# this test should **not** raise a RuntimeWarning
 def test_gh8904_zeroder_at_root_fails():
     """Test that Newton or Halley don't warn if zero derivative at root"""
 
     # a function that has a zero derivative at it's root
     def f_zeroder_root(x):
-        return x ** 3 - x ** 2
+        return x**3 - x**2
 
     # should work with secant
     r = zeros.newton(f_zeroder_root, x0=0)
@@ -185,7 +185,7 @@ def test_gh8904_zeroder_at_root_fails():
 
     # 1st derivative
     def fder(x):
-        return 3 * x ** 2 - 2 * x
+        return 3 * x**2 - 2 * x
 
     # should work with newton and halley
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder)

--- a/scipy/optimize/tests/test_zeros.py
+++ b/scipy/optimize/tests/test_zeros.py
@@ -175,15 +175,13 @@ def test_complex_halley():
 def test_gh8904_zeroder_at_root_fails():
     """Test that Newton or Halley don't warn if zero derivative at root"""
 
-    xtol = finfo(float).eps
-
     # a function that has a zero derivative at it's root
     def f_zeroder_root(x):
         return x ** 3 - x ** 2
 
     # should work with secant
     r = zeros.newton(f_zeroder_root, x0=0)
-    assert_allclose(r, 0, atol=xtol)
+    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
 
     # 1st derivative
     def fder(x):
@@ -191,10 +189,10 @@ def test_gh8904_zeroder_at_root_fails():
 
     # should work with newton and halley
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder)
-    assert_allclose(r, 0, atol=xtol)
+    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
     r = zeros.newton(f_zeroder_root, x0=0, fprime=fder,
                      fprime2=lambda x: 6 * x - 2)
-    assert_allclose(r, 0, atol=xtol)
+    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
 
     # also test that if a root is found we do not raise RuntimeWarning even if
     # the derivative is zero, EG: at x = 0.5, then fval = -0.125 and
@@ -202,7 +200,7 @@ def test_gh8904_zeroder_at_root_fails():
     # root, but if the solver continued with that guess, then it will calculate
     # a zero derivative, so it should return the root w/o RuntimeWarning
     r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder)
-    assert_allclose(r, 0, atol=xtol)
+    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
     r = zeros.newton(f_zeroder_root, x0=0.5, fprime=fder,
                      fprime2=lambda x: 6 * x - 2)
-    assert_allclose(r, 0, atol=xtol)
+    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)

--- a/scipy/optimize/zeros.py
+++ b/scipy/optimize/zeros.py
@@ -3,7 +3,7 @@ from __future__ import division, print_function, absolute_import
 import warnings
 
 from . import _zeros
-from numpy import finfo, sign, sqrt
+from numpy import finfo
 
 _iter = 100
 _xtol = 2e-12
@@ -195,7 +195,7 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
             fval = func(p0, *args)
             funcalls += 1
             # If a root has been found within some tolerance, then terminate
-            if fval == 0 or abs(fval) < _xtol + _rtol * abs(fval):
+            if fval == 0 or abs(fval) < finfo(float).eps:
                 return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
             fder = fprime(p0, *args)
             funcalls += 1

--- a/scipy/optimize/zeros.py
+++ b/scipy/optimize/zeros.py
@@ -195,7 +195,7 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
             fval = func(p0, *args)
             funcalls += 1
             # If a root has been found within some tolerance, then terminate
-            if fval == 0 or abs(fval) < finfo(float).eps:
+            if fval == 0 or abs(fval) < _xtol + _rtol * abs(fval):
                 return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
             fder = fprime(p0, *args)
             funcalls += 1

--- a/scipy/optimize/zeros.py
+++ b/scipy/optimize/zeros.py
@@ -187,21 +187,22 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
     # so it still works if x0 is complex.
     p0 = 1.0 * x0
-    # check if initial guess is the solution
-    if func(p0, *args) == 0:
-        return _results_select(full_output, (p0, 0, 0, _ECONVERGED))
     funcalls = 0
     if fprime is not None:
         # Newton-Raphson method
         for itr in range(maxiter):
+            # first evaluate fval
+            fval = func(p0, *args)
+            funcalls += 1
+            # If a root has been found within some tolerance, then terminate
+            if fval == 0 or abs(fval) < _xtol + _rtol * abs(fval):
+                return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
             fder = fprime(p0, *args)
             funcalls += 1
             if fder == 0:
                 msg = "derivative was zero."
                 warnings.warn(msg, RuntimeWarning)
                 return _results_select(full_output, (p0, funcalls, itr + 1, _ECONVERR))
-            fval = func(p0, *args)
-            funcalls += 1
             newton_step = fval / fder
             if fprime2 is None:
                 # Newton step

--- a/scipy/optimize/zeros.py
+++ b/scipy/optimize/zeros.py
@@ -194,8 +194,8 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
             # first evaluate fval
             fval = func(p0, *args)
             funcalls += 1
-            # If a root has been found within some tolerance, then terminate
-            if fval == 0 or abs(fval) < _xtol + _rtol * abs(fval):
+            # If fval is 0, a root has been found, then terminate
+            if fval == 0:
                 return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
             fder = fprime(p0, *args)
             funcalls += 1

--- a/scipy/optimize/zeros.py
+++ b/scipy/optimize/zeros.py
@@ -187,6 +187,9 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
     # so it still works if x0 is complex.
     p0 = 1.0 * x0
+    # check if initial guess is the solution
+    if func(p0, *args) == 0:
+        return _results_select(full_output, (p0, 0, 0, _ECONVERGED))
     funcalls = 0
     if fprime is not None:
         # Newton-Raphson method

